# ChatGPT Mental Support simulation

In this notebook we will simulate how chatGPT will respond to request from people with mental issues, problems and worries, that will be expressed plain in conversation.
This responses will be compared to responses from mental counselor to see how is chatGPT doing this.

### Dataset

This simulation is based on this dataset:

[Mental health Counseling Conversations dataset](https://www.kaggle.com/datasets/thedevastator/nlp-mental-health-conversations)

### Dataset structure

| Column   | Description | 
|----------|----------|
| Context  | Request, phrase, query sentence from the patient, that seems to have mental troubles, problem, disorders of different kind        |  
| Response | Response, that mental counselor is supposed to give. For one ```Context``` there is from 1 to 47 different responses (mean 2.5) | 


### Comparison

Comparison between ChatGPT response will be done with special prompt, resulting in score [0-100], and be done comparing one response from ChatGRT to all (one by one) responses from the dataset. New dataset will chatGPT responses and scores will be build, and then explored



## Preparing the data

The dataset on the link above is available to download, so here it is already present in the ```dataset/NLP_Mental_Health_Conversations.csv```

In [2]:
#!pip install pandas

In [3]:
import pandas as pd

C:\Users\sirius\AppData\Local\Temp\ipykernel_1380\4080736814.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [4]:
CONVERSATIONS_DATASET_FILENAME='dataset/NLP_Mental_Health_Conversations.csv'

conversation_dataset=pd.read_csv(CONVERSATIONS_DATASET_FILENAME)

conversation_dataset.head()

Context  \
0  I'm going through some things with my feelings...   
1  I'm going through some things with my feelings...   
2  I'm going through some things with my feelings...   
3  I'm going through some things with my feelings...   
4  I'm going through some things with my feelings...   

                                            Response  
0  If everyone thinks you're worthless, then mayb...  
1  Hello, and thank you for your question and see...  
2  First thing I'd suggest is getting the sleep y...  
3  Therapy is essential for those that are feelin...  
4  I first want to let you know that you are not ...

In [5]:
len(conversation_dataset)

3512

In [6]:
conversation_dataset.columns = conversation_dataset.columns.str.lower().str.replace(' ', '_')

In [7]:
conversation_dataset.isna().sum()

context     0
response    4
dtype: int64

In [8]:
conversation_dataset.describe()

context  \
count                                                3512   
unique                                                995   
top     I have so many issues to address. I have a his...   
freq                                                   94   

                                                 response  
count                                                3508  
unique                                               2471  
top     Counseling ends when the client has received t...  
freq                                                    3

In [9]:
context_example=conversation_dataset.loc[328]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)


I'm in a state of depression right now. Who can I talk to? I've been sick in a lot of pain and crying. Don't know where to turn.


In [10]:
responses_example

context  \
328   I'm in a state of depression right now. Who ca...   
2166  I'm in a state of depression right now. Who ca...   

                                               response  
328   First of all, if you’re sick and in pain, have...  
2166  First of all, if you’re sick and in pain, have...

In [11]:
responses_example.reset_index(inplace=True)
responses_example
print(responses_example.loc[0]['response'])
print(responses_example.loc[1]['response'])

First of all, if you’re sick and in pain, have you seen your doctor? If you have a chronic illness and pain, this can often lead to depression and is very common. You’ve taken the first step to dealing with it. You can talk to your doctor, even if your depression isn’t related to your health. He or she can either look at prescribing you an antidepressant or can refer you to a psychologist or psychiatrist.If that’s not really an option for you, there are also helplines (local or national) or support groups where people dealing with similar issues can talk to each other. If you can't find any in your area (newspapers usually publish this information), there are online support groups as well.
First of all, if you’re sick and in pain, have you seen your doctor? If you have a chronic illness and pain, this can often lead to depression and is very common. You’ve taken the first step to dealing with it. You can talk to your doctor, even if your depression isn’t related to your health. He or s

### Dataset contains duplicate responses

lets see some more

In [12]:
context_example=conversation_dataset.loc[38]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I have so many issues to address. I have a history of sexual abuse, I’m a breast cancer survivor and I am a lifetime insomniac.    I have a long history of depression and I’m beginning to have anxiety. I have low self esteem but I’ve been happily married for almost 35 years.
   I’ve never had counseling about any of this. Do I have too many issues to address in counseling?


In [13]:
responses_example

context  \
23    I have so many issues to address. I have a his...   
24    I have so many issues to address. I have a his...   
25    I have so many issues to address. I have a his...   
26    I have so many issues to address. I have a his...   
27    I have so many issues to address. I have a his...   
...                                                 ...   
2010  I have so many issues to address. I have a his...   
2011  I have so many issues to address. I have a his...   
2012  I have so many issues to address. I have a his...   
2013  I have so many issues to address. I have a his...   
2014  I have so many issues to address. I have a his...   

                                               response  
23    Let me start by saying there are never too man...  
24    It is never too late to get help and begin mak...  
25    You have been through so much and it sounds li...  
26    Absolutely not.  I strongly recommending worki...  
27    Absolutely not!  In fact, most people have man...  
...                                                 ...  
2010  The way it goes is that all your issues are co...  
2011  I think this is a very common question that pe...  
2012  Hello, I'm so glad you decided to take the fir...  
2013  Not at all my dear. Human beings are complex c...  
2014  Everyone has topics to discuss when they start...  

[94 rows x 2 columns]

In [14]:
context_example=conversation_dataset.loc[1000]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I'm fine when we start becoming intimate, but out of nowhere, I will get a flashback of what happened to me in the past. I start hysterically crying and freaking out when my boyfriend obviously has done nothing to hurt me.


In [15]:
responses_example

context  \
999   I'm fine when we start becoming intimate, but ...   
1000  I'm fine when we start becoming intimate, but ...   
2701  I'm fine when we start becoming intimate, but ...   
2709  I'm fine when we start becoming intimate, but ...   

                                               response  
999   Have you sought the support of a therapist rea...  
1000  Sexual intimacy can be very triggering for sur...  
2701  Have you sought the support of a therapist rea...  
2709   Sexual intimacy can be very triggering for su...

### Removing duplicates and zero entries

In [16]:
conversation_dataset.dropna(subset=['response'], inplace=True)
conversation_dataset.drop_duplicates(subset=['response'], inplace=True)
conversation_dataset.reset_index(drop=True, inplace=True)

In [17]:
len(conversation_dataset)

2471

In [18]:
conversation_dataset.isna().sum()

context     0
response    0
dtype: int64

In [19]:
context_example=conversation_dataset.loc[328]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)


I'm in a state of depression right now. Who can I talk to? I've been sick in a lot of pain and crying. Don't know where to turn.


In [20]:
responses_example

context  \
328   I'm in a state of depression right now. Who ca...   
2025  I'm in a state of depression right now. Who ca...   

                                               response  
328   First of all, if you’re sick and in pain, have...  
2025  First of all, if you’re sick and in pain, have...

In [21]:
conversation_dataset.describe()

context  \
count                                                2471   
unique                                                912   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2471  
unique                                               2471  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

In [22]:
context_example=conversation_dataset.loc[1000]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I'm fine when we start becoming intimate, but out of nowhere, I will get a flashback of what happened to me in the past. I start hysterically crying and freaking out when my boyfriend obviously has done nothing to hurt me.


In [23]:
responses_example

context  \
999   I'm fine when we start becoming intimate, but ...   
1000  I'm fine when we start becoming intimate, but ...   
2218  I'm fine when we start becoming intimate, but ...   
2223  I'm fine when we start becoming intimate, but ...   

                                               response  
999   Have you sought the support of a therapist rea...  
1000  Sexual intimacy can be very triggering for sur...  
2218  Have you sought the support of a therapist rea...  
2223   Sexual intimacy can be very triggering for su...

In [24]:
conversation_dataset.loc[1000]['response']==conversation_dataset.loc[2226]['response']

False

In [25]:
print(conversation_dataset.loc[1000]['response'])
print(conversation_dataset.loc[2226]['response'])

print(conversation_dataset.loc[328]['response'])
print(conversation_dataset.loc[2025]['response'])

print(conversation_dataset.loc[328]['response']==conversation_dataset.loc[2025]['response'])

Sexual intimacy can be very triggering for survivors even when it is both wanted and consensual.  You may want to consider seeing a therapist who specializes in trauma to work through the abuse if you have not already done so. Often times triggers still hold such a powerful effect when the emotions related to the abuse  have not been fully processed.  In the meantime, you may want to consider coming up with a Safe Word to let your partner know that you are being triggered or to communicate your physical boundaries to him.  Often times, the experience of communicating  your physical boundaries to your partner, having those boundaries respected and validated, and having a partner who is understanding and  willing to engage in intimacy in such a way that does not violate your physical boundaries  can reinforce a sense of safety with him.
EMDR therapy has shown great results for work with PTSD symptoms which are similar to what you described. If you are unable to find a local EMDR speciali

In [26]:
response1 = conversation_dataset.loc[328]['response']
response2 = conversation_dataset.loc[2025]['response']

# Find the minimum length of the two responses
min_length = min(len(response1), len(response2))

# Iterate through each character and find the positions where they differ
differences = []
for i in range(min_length):
    if response1[i] != response2[i]:
        differences.append(i)

# Print the positions where the responses differ
print("Positions where responses differ:")
print(differences)
print('Lengths of response1 and response2:', len(response1), len(response2))

Positions where responses differ:
[]
Lengths of response1 and response2: 698 699


In [27]:
print(response1.strip()==response2.strip())

True


So there is duplicate entries, but they differ in heading or trailing spaces.
Let's fix this, and remove duplicates once more.

In [28]:
conversation_dataset['response']=conversation_dataset['response'].apply(lambda x: x.strip())

In [29]:
conversation_dataset.describe()

context  \
count                                                2471   
unique                                                912   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2471  
unique                                               2052  
top     That sounds like a very hurtful situation. Unf...  
freq                                                    2

In [30]:
conversation_dataset.drop_duplicates(subset=['response'], inplace=True)
conversation_dataset.reset_index(drop=True, inplace=True)

In [31]:
conversation_dataset.describe()

context  \
count                                                2052   
unique                                                843   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2052  
unique                                               2052  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

In [32]:
# Test once more
context_example=conversation_dataset.loc[1000]['context']

responses_example=conversation_dataset[conversation_dataset['context']==context_example]

print(context_example)

I'm fine when we start becoming intimate, but out of nowhere, I will get a flashback of what happened to me in the past. I start hysterically crying and freaking out when my boyfriend obviously has done nothing to hurt me.


In [33]:
responses_example

context  \
999   I'm fine when we start becoming intimate, but ...   
1000  I'm fine when we start becoming intimate, but ...   

                                               response  
999   Have you sought the support of a therapist rea...  
1000  Sexual intimacy can be very triggering for sur...

### Errors in dataset

Working with the dataset I found some irrelevant entries. So we should just delete them


In [34]:
error_context=[" such as not enough sleep", 
"Can a counselor take sides with one parent and allow a parent to order the child to tell the counselor"]

conversation_dataset=conversation_dataset[~conversation_dataset['context'].str.startswith(tuple(error_context))]


In [35]:
conversation_dataset.describe()

context  \
count                                                2049   
unique                                                841   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2049  
unique                                               2049  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

In [36]:
context_unique = conversation_dataset.groupby('context').size().reset_index(name='count')

In [37]:
context_unique.sort_values('count',ascending=False)

context  count
286  I have so many issues to address. I have a his...     47
78   How do you know you have the right therapist f...     41
545  Is it normal for people to cry during therapy,...     38
523  I've gone to a couple therapy sessions so far ...     26
828  When my daughter is stressed about a silly thi...     23
..                                                 ...    ...
360  I stress over everything. If I don't have enou...      1
365  I think I have depression, anxiety, bipolar di...      1
368  I think adult life is making him depressed and...      1
369  I think i may suffer from depression, and it i...      1
840  i was addicted to porn since the second grade....      1

[841 rows x 2 columns]

In [38]:
context_unique.sort_values('count',ascending=False).head(30)

context  count
286  I have so many issues to address. I have a his...     47
78   How do you know you have the right therapist f...     41
545  Is it normal for people to cry during therapy,...     38
523  I've gone to a couple therapy sessions so far ...     26
828  When my daughter is stressed about a silly thi...     23
461  I'm going through some things with my feelings...     23
80     How does a person start the counseling process?     22
79   How does a counselor decide when to end counse...     20
31   Does counseling really do anything that can he...     20
73   How do I decide if I should end my relationshi...     15
41   Every winter I find myself getting sad because...     15
695  My mother takes care of niece whom my sister a...     13
353  I start counseling/therapy in a few days (I'm ...     12
306  I just took a job that requires me to travel f...     11
396  I was raped a couple months ago, Since then, a...     11
823  When I'm in large crowds I get angry and I jus...     11
643  My fiancé and I have been together for 3 years...     11
311  I keep having these random thoughts that I don...     11
213  I have a friend that who I used to be in a rel...     10
693  My mother has Alzheimer's and she has become s...     10
484  I'm obsessing about a terrible breakup. Everyt...      9
624  My daughter is in later elementary school. She...      9
303  I just lost my grandpa and i'm having a rough ...      9
465  I'm having relationship problems and I want to...      9
813                What makes a healthy marriage last?      9
154  I don't know what to say. I have never really ...      9
566  I’ve been in a long distance relationship with...      9
246  I have been married for 11 years. Within the p...      9
16   After he got home from the hospital he was ang...      9
561  I’m facing severe depression and anxiety and I...      9

In [39]:
context_unique['count'].describe()

count    841.000000
mean       2.436385
std        3.538705
min        1.000000
25%        1.000000
50%        1.000000
75%        2.000000
max       47.000000
Name: count, dtype: float64

In [40]:
context_unique[context_unique['count']==1]

context  count
0    A few nights ago I talked to this girl I know ...      1
2    A friend of mine taking psychology advised I g...      1
3    A girl and I were madly in love. We dated for ...      1
8    About a month ago, I went through my boyfriend...      1
13   After a domestic abuse situation, I went to a ...      1
..                                                 ...    ...
826  When having sex I think of other men and want ...      1
833  Whenever I leave my girlfriend I get panic att...      1
835  Whenever I sing along with a song that I like ...      1
837                   Why am I attracted to older men?      1
840  i was addicted to porn since the second grade....      1

[452 rows x 2 columns]

In [41]:
count_distribution = context_unique['count'].value_counts()

count_distribution

count
1     452
2     182
3      78
4      40
5      23
6      18
9      12
7      10
8       6
11      5
23      2
20      2
15      2
10      2
38      1
26      1
41      1
12      1
47      1
22      1
13      1
Name: count, dtype: int64

In [41]:
CONTEXT_UNIQUE_FILENAME='working_data/context_unique.csv'

context_unique.to_csv(CONTEXT_UNIQUE_FILENAME)

In [42]:
conversation_dataset.describe()

context  \
count                                                2049   
unique                                                841   
top     I have so many issues to address. I have a his...   
freq                                                   47   

                                                 response  
count                                                2049  
unique                                               2049  
top     If everyone thinks you're worthless, then mayb...  
freq                                                    1

### Data is cleaned and ready for simulation

So we have 845 unique requests, with a mean of 2.45 responses for each, and 452 having one response, and max of 47 responses, but only 10 requests have more then 10 responses.

### Conclusions

**Maybe for simulation we should use only requests with 2 or more responses, so we have some variety in approved responses to compare with. If we take request with 2 or more we will have 393 requests, if we take with 3 or more - 209 requests. 209 may be enough for our simulation. And having only one response will make it possible for incorrect scoring of ChatGPT output. Because mental issues is delicate matter, so several different responses may serve well for specific case.**


## ChatGPT

We follow the standard way of using OpenAI API for reaching ChatGPT
The below code uses ```.env``` file in this directory, which is not included to the repository.
```.env```  which must contain:
```
OPENAI_API_KEY=...
```
This API key to be used to interact with OpenAI API

If you want to use this code, you must [get your own API key from OpenAI](https://platform.openai.com/api-keys)

Moreover you will need to make a payment to your ChatGPT account, to be able to use it.

This code is based on the [course by DeepLearning.AI](https://learn.deeplearning.ai/courses/langchain/lesson/2/models%2C-prompts-and-parsers)

In [43]:
# !pip install python-dotenv
# !pip install openai

In [61]:
import os
import openai

from dotenv import load_dotenv, find_dotenv
_ = load_dotenv(find_dotenv()) # read local .env file

In [55]:
GPT_MODEL='gpt-3.5-turbo-0125'   
#
# Pricing
# Model	Input	Output
# gpt-3.5-turbo-0125	$0.0005 / 1K tokens	$0.0015 / 1K tokens
#
# Updated GPT 3.5 Turbo
# The latest GPT-3.5 Turbo model with higher accuracy at responding in requested formats 
# and a fix for a bug which caused a text encoding issue for non-English language function calls. 
# Returns a maximum of 4,096 output tokens.
#

In [62]:
client = openai.OpenAI(

    api_key=os.environ.get("OPENAI_API_KEY"),
)

In [47]:
chat_completion = client.chat.completions.create(
    messages=[
        {
            "role": "user",
            "content": "Hello, this is a test",
        }
    ],
    model=GPT_MODEL,
)

# print(chat_completion.choices[0].message.content)


In [53]:
import random

TEMPERATURE = 0.6 # default temperature for ChatGPT. Web ChatGPT uses 0.6 all the time. varies from 0 (same answers, no randomness) to 1 (much variety, much random)

# Our randomness constants:

USE_STARTUP_DIALOG = 0.25

# if we use startup dialog, we may have up to 4 preceding phrases.

DIALOG_LENGTH_WEIGHTS=[0.35,0.3,0.2,0.15]  # So len 1 will occur more frequently, len 4 - much less
DIALOG_LENGTH_OPTIONS=[2,4,6,8]

starting_messages = [
[
    {'role': 'user', 'content': "Hi"},
    {'role': 'assistant', 'content': "Hello! How can I assist you today?"},
    {'role': 'user', 'content': "Yes! How are you?"},
    {'role': 'assistant', 'content': "As an AI, I don't have feelings or emotions, but I'm here and ready to assist you with any questions or tasks you have. How can I help you today?"},
    {'role': 'user', 'content': "What do you do"},
    {'role': 'assistant', 'content': "I'm here to help with a wide range of tasks! You can ask me questions, seek information on various topics, brainstorm ideas, get assistance with writing or editing, play text-based games, generate creative content, and much more. Just let me know what you need help with, and I'll do my best to assist you!"},
    {'role': 'user', 'content': 'Yes'},
    {'role': 'assistant', 'content': "Great! If you have any specific questions or need assistance with something, feel free to ask. I'm here to help!"},
],
[
    {'role': 'user', 'content': "Hello!"},
    {'role': 'assistant', 'content': "Hello! How can I assist you today?"},
    {'role': 'user', 'content': "I don't know"},
    {'role': 'assistant', 'content': "That's perfectly fine! If you have any questions, need advice, want to discuss a topic, or just want to chat, feel free to let me know. I'm here to help with anything you need."},
    {'role': 'user', 'content': "Ok. I will think about it"},
    {'role': 'assistant', 'content': "Sure thing! Take your time to think, and whenever you're ready, feel free to reach out. I'll be here to assist you whenever you need."},
    {'role': 'user', 'content': "Tell me something"},
    {'role': 'assistant', 'content': "Sure! Did you know that the world's largest desert isn't the Sahara, but actually Antarctica? Despite being covered in ice, Antarctica is considered a desert because it receives very little precipitation, less than 2 inches (50 millimeters) per year in some areas. Its size and harsh climate make it the largest desert on Earth, even larger than the Sahara Desert."},         
],
[
    {'role': 'user', 'content': "Nice to see you"},
    {'role': 'assistant', 'content': "Nice to see you too! How can I assist you today?"},
    {'role': 'user', 'content': "What can you say?"},
    {'role': 'assistant', 'content': "As an AI language model, I can engage in conversation, answer questions, provide information on a wide range of topics, generate text based on prompts, offer suggestions, and more. Just let me know what you need assistance with or what you'd like to discuss!"},
    {'role': 'user', 'content': "Just tell me something"},
    {'role': 'assistant', 'content': "Sure! Did you know that honey never spoils? Archaeologists have found pots of honey in ancient Egyptian tombs that are over 3,000 years old and still perfectly edible. This remarkable longevity is due to the unique chemical composition of honey, which includes low water content and natural acidity, making it inhospitable to bacteria and microorganisms that cause food to spoil."},
    {'role': 'user', 'content': "Interesting"},
    {'role': 'assistant', 'content': "I'm glad you found it interesting! If you have any other topics you'd like to learn about or discuss, feel free to let me know. I'm here to help!"},  
],
[
    {'role': 'user', 'content': "How are things"},
    {'role': 'assistant', 'content': "As an AI, I don't have personal experiences, but I'm here and ready to assist you with any questions or tasks you have! How can I help you today?"},  
    {'role': 'user', 'content': "I am unsure"},
    {'role': 'assistant', 'content': "It's okay to feel unsure sometimes. If there's anything specific on your mind or any decision you're facing, feel free to share, and I'll do my best to assist you. Whether it's about making a choice, seeking advice, or just discussing your thoughts, I'm here to help."},  
    {'role': 'user', 'content': "yes talk to me"},
    {'role': 'assistant', 'content': "Of course! Is there anything particular on your mind that you'd like to talk about? It could be anything from personal concerns, questions about a topic, or just casual conversation. Feel free to share, and I'm here to listen and offer assistance or insights as best as I can."},  
    {'role': 'user', 'content': "i don't know"},
    {'role': 'assistant', 'content': "That's completely fine. Sometimes it can be helpful to explore different topics or interests to see if anything sparks your curiosity or engages your thoughts. Is there a subject you've been curious about or something you've been meaning to learn more about? Alternatively, we can engage in a casual conversation about anything you'd like. Let me know if there's a specific direction you'd like to take our conversation or if there's anything specific you'd like to discuss."},  
],
]


### Helper function for request - response

In [49]:
def generate_messages(phrase_request:str)->list[dict]:
    """Using random this function will generate messages to pass to ChatGPT

    Args:
        phrase_request (str): Request phrase to put to the end of messages list

    Returns:
        list[dict]: messages list in a OpenAi format
    """
    # this is a standard option
    return_messages=[    {'role': 'system', 'content': "You are a helpful assistant."},
                     ]
    
    # use random to chose if we are adding anything
    if random.random()<USE_STARTUP_DIALOG:
        # choose number of adding, based on provided weights
        startup_len_to_add=random.choices(DIALOG_LENGTH_OPTIONS, weights=DIALOG_LENGTH_WEIGHTS)[0]
        message_chain=random.randint(0,len(starting_messages)-1)
        return_messages+=starting_messages[message_chain][:startup_len_to_add]
    
    # finally our phrase                                                                                
    return return_messages+[
        {'role': 'user', 'content': phrase_request},
    ]

In [50]:
generate_messages('hello')

[{'role': 'system', 'content': 'You are a helpful assistant.'},
 {'role': 'user', 'content': 'hello'}]

In [51]:
generate_messages('hello')

[{'role': 'system', 'content': 'You are a helpful assistant.'},
 {'role': 'user', 'content': 'Hi'},
 {'role': 'assistant', 'content': 'Hello! How can I assist you today?'},
 {'role': 'user', 'content': 'Yes! How are you?'},
 {'role': 'assistant',
  'content': "As an AI, I don't have feelings or emotions, but I'm here and ready to assist you with any questions or tasks you have. How can I help you today?"},
 {'role': 'user', 'content': 'What do you do'},
 {'role': 'assistant',
  'content': "I'm here to help with a wide range of tasks! You can ask me questions, seek information on various topics, brainstorm ideas, get assistance with writing or editing, play text-based games, generate creative content, and much more. Just let me know what you need help with, and I'll do my best to assist you!"},
 {'role': 'user', 'content': 'Yes'},
 {'role': 'assistant',
  'content': "Great! If you have any specific questions or need assistance with something, feel free to ask. I'm here to help!"},
 {'ro

In [52]:
# now let's write a function that sends this messages to ChatGPT and get response

In [56]:
def request_gpt(messages:list[dict], temperature=TEMPERATURE, model=GPT_MODEL)->str:
    """Sends messages to ChatGPT and returns the response

    Args:
        messages (list[dict]): list of messages in OpenAI format

    Returns:
        str: response of ChatGpt
    """
    chat_completion = client.chat.completions.create(
        messages=messages,
        model=model,
        temperature=temperature
    )

    return chat_completion.choices[0].message.content

In [54]:
def speak_with_gpt(phrase_request:str)->str:
    """Use phrase_request to generate messages with generate_messages function, send's them to ChatGPT, and return response

    Args:
        phrase_request (str): Phrase to ask

    Returns:
        str: Response of ChatGPT
    """
    return request_gpt(generate_messages(phrase_request=phrase_request))

In [55]:
example_phrase=context_unique.loc[100]['context']

In [56]:
generate_messages(example_phrase)

[{'role': 'system', 'content': 'You are a helpful assistant.'},
 {'role': 'user',
  'content': 'I am always arguing with my father. He gets stressed over work and health and talks to me in a tone of voice that seems very demanding and seems more like yelling. I get upset often at this. Am I too sensitive? He always says I am overreacting but never seems to understand that he is hurting my feelings. No matter how often I try to tell him this, he never listens.'}]

In [57]:
# speak_with_gpt(example_phrase)

In [58]:
# lets try it again several times

In [59]:
# speak_with_gpt(example_phrase)

In [60]:
# speak_with_gpt(example_phrase)

In [61]:
# copies of output
chatgpt_response1="It's completely understandable that you feel upset when your father speaks to you in a demanding or yelling tone. It's important to remember that everyone has different communication styles and ways of expressing their stress. It's also okay to have your own feelings and reactions to the way he speaks to you.\n\nIt's not necessarily that you are too sensitive, but rather that you have valid feelings that deserve to be acknowledged and respected. It's important to communicate your feelings to your father in a calm and assertive manner. You can try to have a conversation with him during a calm moment, expressing how his tone of voice makes you feel and how you would like to be spoken to.\n\nIt's possible that your father may not be aware of how his tone affects you, and it may take some time for him to understand and adjust his communication style. It's important to be patient and persistent in expressing your feelings to him.\n\nIf despite your efforts, your father continues to dismiss your feelings and refuses to listen, it may be helpful to seek support from a trusted family member, friend, or counselor to help navigate the situation and find healthy ways to cope with the challenging interactions. Remember that it's okay to prioritize your own well-being and set boundaries in relationships that are emotionally challenging."
chatgpt_response2='It sounds like you\'re going through a challenging situation with your father. It\'s important to remember that everyone experiences and expresses stress differently. It\'s also common for misunderstandings to occur in communication, especially during times of tension.\n\nYour feelings are valid, and it\'s okay to feel upset when someone speaks to you in a way that hurts your feelings. It\'s also understandable that you want your father to understand how his tone affects you.\n\nHere are a few suggestions that may help improve communication with your father:\n\n1. Choose the right time: Try to have a calm and private conversation with your father when both of you are in a relaxed state and not already in the midst of an argument.\n\n2. Use "I" statements: Express your feelings using "I" statements to help your father understand how his tone of voice affects you. For example, "I feel hurt when you speak to me in a demanding tone."\n\n3. Listen actively: Try to listen to your father\'s perspective as well. Understanding each other\'s point of view can lead to better communication and empathy.\n\n4. Seek support: If you find it challenging to communicate with your father on your own, consider seeking the help of a family therapist or counselor who can facilitate productive conversations and provide guidance on improving your relationship.\n\nRemember, it\'s okay to set boundaries and communicate your feelings in a respectful manner. Your well-being is important, and it\'s essential to address any concerns that are affecting your mental health and relationships.'
chatgpt_response3="It's understandable that you feel upset in this situation. It's important to remember that your feelings are valid, and it's not necessarily about being too sensitive. It sounds like there may be a communication issue between you and your father.\n\nIt might be helpful to have a calm and honest conversation with your father when you both are in a good mood. Express how his tone of voice and demanding behavior make you feel, and try to explain that you are not trying to overreact but instead want to improve your relationship with him.\n\nIt's possible that your father may not realize the impact of his words and actions on you, so having an open and honest conversation could help bring this to his attention. It may take time for him to understand and change his behavior, so be patient and continue to communicate your feelings to him.\n\nIn the meantime, it could also be beneficial to find healthy ways to cope with the stress and tension in your relationship with your father, such as practicing self-care, talking to a trusted friend or counselor, or engaging in activities that bring you joy and relaxation. Remember to prioritize your well-being and set boundaries to protect your emotional health."


In [106]:
dataset_response=conversation_dataset[conversation_dataset['context']==example_phrase]['response'].iloc[0]
print(dataset_response)

Thanks for the question. Regardless of whether or not you are sensitive, it is okay to want better communication with those who are important to you. When we feel that we are being talked down to or being ridiculed, it is difficult to not take it personally. It sounds like you have taken the first and necessary step in resolving this by bringing up the issue to him. Is it possible that the delivery of your message is causing him to feel blamed or defensive? Try communicating with him during a time when you are not upset and when it is out of context. Begin your statements with "I" as opposed to "you." For example, you can tell him, "I feel angry when you raise your voice at me" instead of "You always yell at me." Also, avoid using black and white terms such as never and always. Additionally, offer sugggestions as to ways that he can better communicate with you. Maybe he just doesn't know how to. Lastly, when delivering a message, I like to use the sandwich method by starting off the di

### Helper function for comparing the results

In [63]:
COMPARING_TEMPERATURE=0.3   # lower is better in such cases, we will get more sustainable responses

#
# note: i have made several experiments with it. with high temp you sometimes have 20 f
#

COMPARING_PROMPT_TEMPLATE_1 ="""
Please compare the meaning and summarization of PHRASE_1 and PHRASE_2. 

Rate similarity of meaning by scale from 0 to 100 (0 - opposite meaning, no similarity, 100 - total similarity).

Response with number only!

```PHRASE_1```
<<<PHRASE1###>>>
```END OF PHRASE_1```


```PHRASE_2```
<<<PHRASE2###>>>
```END OF PHRASE_2```

SCALE OF LIKENESS:
"""

COMPARING_PROMPT_TEMPLATE_2 ="""
You will need to compare two phrases and scale their likeness. 

Compare PHRASE_1 and PHRASE_2 by summarized meaning, rate similarity by scale from 0 to 100 (0 - no similarity or opposite meaning, 100 - total likeness, similar meaning).

Response with number only!


```PHRASE_1```
<<<PHRASE1###>>>
```END OF PHRASE_1```


```PHRASE_2```
<<<PHRASE2###>>>
```END OF PHRASE_2```

SCALE OF LIKENESS:
"""

In [64]:
def generate_comparing_prompt(response1:str,response2:str, comparing_template=COMPARING_PROMPT_TEMPLATE_2)->str:
    """Insert into COMPARING_PROMPT_TEMPLATE given request and responses

    Args:
        response1 (str): response1
        response2 (str): response2

    Returns:
        str: Prompt to send to ChatGRT for comparing
    """
    return comparing_template.replace('<<<PHRASE1###>>>', response1).replace('<<<PHRASE2###>>>', response2)

In [65]:
comparing_prompt=generate_comparing_prompt(chatgpt_response3,dataset_response)
print(comparing_prompt)


You will need to compare two phrases and scale their likeness. 

Compare PHRASE_1 and PHRASE_2 by summarized meaning, rate similarity by scale from 0 to 100 (0 - no similarity or opposite meaning, 100 - total likeness, similar meaning).

Response with number only!


```PHRASE_1```
It's understandable that you feel upset in this situation. It's important to remember that your feelings are valid, and it's not necessarily about being too sensitive. It sounds like there may be a communication issue between you and your father.

It might be helpful to have a calm and honest conversation with your father when you both are in a good mood. Express how his tone of voice and demanding behavior make you feel, and try to explain that you are not trying to overreact but instead want to improve your relationship with him.

It's possible that your father may not realize the impact of his words and actions on you, so having an open and honest conversation could help bring this to his attention. It ma

In [66]:
def compare_with_gpt(response1:str,response2:str, temperature=COMPARING_TEMPERATURE, model=GPT_MODEL)->int:
    """Makes a prompt to compare phrase1 and phrase2, send to ChatGPT and return and integer 

    Args:
        response1 (str): response1
        response2 (str): response2

    Returns:
        int: 0..100 - scale of likeness, -1 if something goes wrong
    """
    messages=[{'role': 'system', 'content': "You are a helpful assistant. "},
              {'role': 'user', 'content': generate_comparing_prompt(response1,response2)},  
    ]
    response=request_gpt(messages,temperature=temperature)
    
    try:
        scale_number=int(response.strip())
    except ValueError:
        scale_number=-1
        
    return scale_number
    

In [88]:
# print(compare_with_gpt(chatgpt_response1,dataset_response))

90


In [86]:
# print(compare_with_gpt(chatgpt_response2,dataset_response))

85


In [84]:
# print(compare_with_gpt("rtu7e56ujdfghj",dataset_response))

10


In [107]:
# print(compare_with_gpt("Oh no!! You are a bad son!!!",dataset_response))

10


In [92]:
example_phrase2=context_unique.iloc[340]['context']

In [93]:
example_phrase2

'I panicked over a minor parking lot mistake. It was totally my fault, but due to another insurance issue, I left a note with an illegible phone number. Guilty over this, I got an envelope containing several hundred dollars. Unfortunately, by the time I returned, the other driver had seen my ersatz note and drove away. Since I cannot make this up to that individual, how do I atone for my cowardly act? I have been trying to do random acts of kindness in hopes that karma will somehow even out.'

In [94]:
dataset_response=conversation_dataset[conversation_dataset['context']==example_phrase2]['response']
print(dataset_response)

407    From your writing, you have already atoned for...
408    Thanks for sharing as it shows your humanity a...
Name: response, dtype: object


In [95]:
dataset_response1=dataset_response.iloc[0]
dataset_response2=dataset_response.iloc[1]


In [96]:
print(dataset_response1,dataset_response2)

From your writing, you have already atoned for what you describe as "cowardly".Recognizing your action and trying to directly correct it, realizing your action was improper and demonstrating the willingness to take right action, are all examples of making things right.Since the actual driver is unavailable to consult as to their wishes on what they expect and feel would correct the matter, you did the next best thing and considered the situation from the driver point of view. Thanks for sharing as it shows your humanity and that you have a sensitive conscience. In your example here I'd encourage you to always think, "safety first"; and never offer a stranger at the accident scene "...several hundred dollars". Why? Because it could go badly against you (i.e. extortion, robbed, or would this be considered a bribe).   To the point, I'm reminded of Steps Eight and Nine in many 12-step programs regarding making amends. In your specific example since the injured unidentified party departed, 

In [97]:
chat_gpt_responses=[]
print("Example simulation 1.")
print("request:",example_phrase2)
print("database response1:", dataset_response1 )
print("database response2:", dataset_response2 )
print()

def simulation1():
    for i in range(4):
        print('iteration', i)
        single_response=speak_with_gpt(example_phrase2)
        print("GPT have given response", i)
        comp_score1=compare_with_gpt(single_response,dataset_response1)
        comp_score2=compare_with_gpt(single_response,dataset_response2)
        print("Compared to dataset response 1:", comp_score1)
        print("Compared to dataset response 2:", comp_score2)   
        chat_gpt_responses.append([comp_score1,comp_score2,single_response])
        print()
        
# simulation1()
    
# print('final array of comparing:', chat_gpt_responses)

Example simulation 1.
request: I panicked over a minor parking lot mistake. It was totally my fault, but due to another insurance issue, I left a note with an illegible phone number. Guilty over this, I got an envelope containing several hundred dollars. Unfortunately, by the time I returned, the other driver had seen my ersatz note and drove away. Since I cannot make this up to that individual, how do I atone for my cowardly act? I have been trying to do random acts of kindness in hopes that karma will somehow even out.
database response1: From your writing, you have already atoned for what you describe as "cowardly".Recognizing your action and trying to directly correct it, realizing your action was improper and demonstrating the willingness to take right action, are all examples of making things right.Since the actual driver is unavailable to consult as to their wishes on what they expect and feel would correct the matter, you did the next best thing and considered the situation fro

In [98]:
chat_gpt_responses

[[78,
  72,
  "It's understandable that you feel guilty about the situation, but it's important to remember that mistakes happen and it's how we choose to learn and grow from them that truly matters. It's great that you're trying to make amends through random acts of kindness.\n\nTo continue atoning for your actions, you can consider the following steps:\n\n1. **Pay it forward:** Continue with your random acts of kindness and spread positivity wherever you go. Small gestures, like helping someone carry groceries or holding the door open, can make a big difference.\n\n2. **Volunteer:** Consider volunteering your time and skills to help those in need. It could be at a local charity, community organization, or even just helping out a neighbor.\n\n3. **Donate:** If you're able to, consider making a donation to a charity or cause that you care about. Every little bit helps and can make a positive impact.\n\n4. **Practice mindfulness:** Reflect on the situation and use it as a learning exper

In [99]:
example_phrase3=context_unique.iloc[1]['context']
dataset_response=conversation_dataset[conversation_dataset['context']==example_phrase3]['response']
print(example_phrase3)
print(dataset_response)

A few years ago I was making love to my wife when for no known reason I lost my erection,    Now I'm In my early 30s and my problem has become more and more frequent. This is causing major problems for my ego and it's diminishing my self esteem. This has resulted in ongoing depression and tearing apart my marriage.    I am devastated and cannot find a cause for these issues. I am very attracted to my wife and want to express it in the bedroom like I used to.    What could be causing this, and what can I do about it?
89    First step always is to do a medical rule out ...
90    When I'm working with men with this type of si...
91    For starters, know that this is a normal exper...
92    Erection maintenance is caused by connective t...
93    If you haven't already, please see a doctor.  ...
94    Assuming there are no medical explanations for...
Name: response, dtype: object


In [100]:
chat_gpt_responses2=[]
print("Example simulation 2.")
print("request:",example_phrase3)

print()

def simulation2():
    for i in range(5):
        print('iteration', i)
        single_response=speak_with_gpt(example_phrase3)
        print("Got GPT response", i)
        result_line=[]
        for j in range(6):
            result_line.append(compare_with_gpt(single_response,dataset_response.iloc[j])) 
        print('comparing scores:',result_line)
        result_line.append(single_response)
        chat_gpt_responses2.append(result_line)
        print()
        
# simulation2()

Example simulation 2.
request: A few years ago I was making love to my wife when for no known reason I lost my erection,    Now I'm In my early 30s and my problem has become more and more frequent. This is causing major problems for my ego and it's diminishing my self esteem. This has resulted in ongoing depression and tearing apart my marriage.    I am devastated and cannot find a cause for these issues. I am very attracted to my wife and want to express it in the bedroom like I used to.    What could be causing this, and what can I do about it?

iteration 0
Got GPT response 0
comparing scores: [70, 70, 85, 10, 80, 80]

iteration 1
Got GPT response 1
comparing scores: [70, 70, 85, 35, 72, 85]

iteration 2
Got GPT response 2
comparing scores: [70, 70, 85, 40, 80, 85]

iteration 3
Got GPT response 3
comparing scores: [72, 70, 85, 20, 80, 85]

iteration 4
Got GPT response 4
comparing scores: [70, 70, 76, 70, 70, 87]



In [101]:
chat_gpt_responses2

[[70,
  70,
  85,
  10,
  80,
  80,
  "I'm sorry to hear that you're experiencing difficulties with your sexual health, and I appreciate you reaching out for help. Erectile dysfunction can be caused by a variety of physical and psychological factors. Some common causes include stress, anxiety, relationship issues, depression, fatigue, alcohol or drug use, certain medications, and underlying health conditions such as diabetes or heart disease.\n\nIt's important to consult with a healthcare provider to determine the root cause of your erectile dysfunction. They can help identify any underlying medical conditions and provide appropriate treatment options. In some cases, counseling or therapy may also be beneficial to address any psychological factors contributing to the issue.\n\nIn the meantime, here are some general tips that may help improve your sexual health and overall well-being:\n\n1. Communicate openly with your partner about your feelings and concerns. Building a supportive and 

In [102]:
ii=38

example_phrase4=context_unique.iloc[ii]['context']
dataset_response_s4=conversation_dataset[conversation_dataset['context']==example_phrase4]['response']
print(example_phrase4)
print(dataset_response_s4)

Every time I send a message to someone or a group message on Instagram, iMessage, or snapchat people will read my messages but then they won't answer me.    Could it be that there is something they don't like about me? I don't understand why they won't answer my messages. How do I get people to respond to me?
1831    I understand the feeling of not being liked du...
1832    Sorry to hear your friends aren't responding t...
1833    I'm sorry that you are having a difficult time...
Name: response, dtype: object


In [103]:
chat_gpt_responses4=[]
print("Example simulation 3.")
print("request:",example_phrase4)

print()

def simulation3():
    for i in range(5):
        print('iteration', i)
        single_response=speak_with_gpt(example_phrase4)
        print("Got GPT response", i)
        result_line=[]
        for j in range(3):
            result_line.append(compare_with_gpt(single_response,dataset_response_s4.iloc[j])) 
        print('comparing scores:',result_line)
        result_line.append(single_response)
        chat_gpt_responses4.append(result_line)
        print()
        
# simulation3()

Example simulation 3.
request: Every time I send a message to someone or a group message on Instagram, iMessage, or snapchat people will read my messages but then they won't answer me.    Could it be that there is something they don't like about me? I don't understand why they won't answer my messages. How do I get people to respond to me?

iteration 0
Got GPT response 0
comparing scores: [72, 85, 70]

iteration 1
Got GPT response 1
comparing scores: [80, 75, 70]

iteration 2
Got GPT response 2
comparing scores: [87, 70, 70]

iteration 3
Got GPT response 3
comparing scores: [72, 72, 70]

iteration 4
Got GPT response 4
comparing scores: [73, 72, 70]



In [104]:
chat_gpt_responses4

[[72,
  85,
  70,
  "I'm sorry to hear that you're feeling this way. It's important to remember that people's behaviors and responses can be influenced by various factors, and it may not always be about you personally. There could be many reasons why someone doesn't respond to a message promptly, such as being busy, distracted, or simply forgetting to reply.\n\nHere are a few tips that may help improve your chances of getting responses:\n\n1. **Be patient:** Give people some time to respond. Not everyone checks their messages immediately, and they may need some time to formulate a response.\n\n2. **Engage in meaningful conversations:** Try to initiate conversations that are interesting and engaging. Ask questions, share stories, or show genuine interest in the other person's life. This can make your messages more likely to receive a response.\n\n3. **Respect boundaries:** Everyone has different communication preferences and boundaries. Make sure you're not overwhelming people with too 

In [119]:
for j in range(3):
    print(dataset_response_s4.iloc[j])

I understand the feeling of not being liked due to not having any responses on the sites you list.More than likely the non-response has less to do with liking you than w liking the way you write your messages.Have you tried asking a question in your posts?  This would be one way to invite people to answer your post.When you write,  imagine yourself as the reader of the post.If you think about the wording of a question which would motivate you to respond, then this formula will very likely be the same for many others who read your post.Good luck!
Sorry to hear your friends aren't responding to you. If these friends are in-person as well as online, perhaps going to them in person and asking talking to them about your concern. Perhaps it's the way you send messages or the way they are receiving them. if they say no, then a simple request to respond to your messages. Measure the result and notice if there are even small differences. Also understand, why this is so important to you. Ask you

## Labeling 

In [65]:
LABELING_TEMPERATURE=0

LABEL_PROMPT_1="""
Analyze the text and label the topic of it. Label must be of one or several words, and name the type of mental disorder the author of text is facing. 
Examples of topics: [ Depression, Anxiety, Eating Disorder, Drug Use, Addiction, Schizophrenia, Bipolar Disorder, Self-Esteem, Personality Disorders, .... ]
You should use common label names with low variety, so I could use this label later in my project.
Respond with label names only, separate different labels with coma ","!
<<<TEXT>>>
<<<<<<TEXT###>>>>>>
<<<END OF TEXT>>>

<<<LABEL>>>:
"""


def generate_labeling_prompt(text_to_label:str, comparing_template=LABEL_PROMPT_1)->str:
    """Insert into comparing_template given text to send later to ChatGPT

    Args:
        text_to_label (str): text to label

    Returns:
        str: Prompt to send to ChatGRT for labeling
    """
    return comparing_template.replace('<<<<<<TEXT###>>>>>>', text_to_label)



In [66]:
ii=39

example_phrase8=context_unique.iloc[ii]['context']

print(example_phrase8)


Every time I speak, he says I anger him. Also, when he’s mad at something, like work, he gets angry at me. I feel like we’re so distant from each other now. We used to laugh all the time, and I feel like he lets too much negativity in. What should I do? We've been married for two years but together for seven.


In [67]:
print(generate_labeling_prompt(example_phrase8))


Analyze the text and label the topic of it. Label must be of one or several words, and name the type of mental disorder the author of text is facing. 
Examples of topics: [ Depression, Anxiety, Eating Disorder, Drug Use, Addiction, Schizophrenia, Bipolar Disorder, Self-Esteem, Personality Disorders, .... ]
You should use common label names with low variety, so I could use this label later in my project.
Respond with label names only, separate different labels with coma ","!
<<<TEXT>>>
Every time I speak, he says I anger him. Also, when he’s mad at something, like work, he gets angry at me. I feel like we’re so distant from each other now. We used to laugh all the time, and I feel like he lets too much negativity in. What should I do? We've been married for two years but together for seven.
<<<END OF TEXT>>>

<<<LABEL>>>:



In [68]:
def get_label_from_chatGPT(text_to_label:str, temperature=LABELING_TEMPERATURE)->str:
    """Generate label prompt, send to ChatGPT and get label for given text

    Args:
        text_to_label (str): text to label

    Returns:
        str: label
    """
    messages=[{'role': 'system', 'content': "You are a helpful assistant. You label text with labels from a limited common set, so they could later be used to aggregate texts."},
              {'role': 'user', 'content': generate_labeling_prompt(text_to_label)},  
    ]
    response=request_gpt(messages,temperature=temperature)
    
    return response
       

In [69]:
print(get_label_from_chatGPT(example_phrase8))

Relationship Issues, Anger


In [71]:
ii=20

example_phrase9=context_unique.iloc[ii]['context']

print(example_phrase9)

An organization admitted to implanting thoughts in my head with technologies. The study was to implant a fantasy other people are having, but to me, it's a nightmare. I lost my kids, my job, and all that. I was an unwilling participant. I no longer trust a therapist. I'm too afraid to go under hypnosis or anything.


In [72]:
print(get_label_from_chatGPT(example_phrase9))

Paranoia, Trauma


# Simulation

work is continued in the next [simulation notebook](notebook_chatgpt_simulation.ipynb)

